pricegraph/src/orderbook.rs

```diff
@@ -0,0 +1,207 @@
+//! Implementation of a graph representation of an orderbook where tokens are
+//! vertices and orders are edges (with users and balances as auxiliary data
+//! to these edges).
+//!
+//! Storage is optimized for graph-related operations such as listing the edges
+//! (i.e. orders) connecting a token pair.
+
+use crate::encoding::{Element, InvalidLength, Price, TokenId, TokenPair, UserId};
+use petgraph::graph::DiGraph;
+use std::cmp;
+use std::collections::HashMap;
+use std::f64;
+
+/// A graph representation of a complete orderbook.
+#[cfg_attr(test, derive(Debug))]
+pub struct Orderbook {
+    /// A map of sell tokens to a mapping of buy tokens to orders such that
+    /// `orders[sell][buy]` is a vector of orders selling token `sell` and
+    /// buying token `buy`.
+    orders: OrderMap,
+    /// A projection of the order book onto a graph of lowest priced orders
+    /// between tokens.
+    projection: DiGraph<TokenId, f64>,
+}
+
+impl Orderbook {
+    /// Reads an orderbook from encoded bytes returning an error if the encoded
+    /// orders are invalid.
+    pub fn read(bytes: impl AsRef<[u8]>) -> Result<Orderbook, InvalidLength> {
+        let mut max_token = 0;
+        let mut orders = OrderMap::new();
+
+        for element in Element::read_all(bytes.as_ref())? {
+            let TokenPair { buy, sell } = element.pair;
+            max_token = cmp::max(max_token, cmp::max(buy, sell));
+            orders
+                .entry(sell)
+                .or_default()
+                .entry(buy)
+                .or_default()
+                .push(element.into());
+        }
+
+        // NOTE: Sort the orders per token pair in descending order, this makes
+        // it so the last order is the cheapest one and can be determined given
+        // a token pair in `O(1)`, and it can simply be `pop`-ed once its amount
+        // is used up and removed from the graph in `O(1)` as well.
+        for pair_orders in orders.values_mut().flat_map(|o| o.values_mut()) {
+            pair_orders.sort_unstable_by(Order::cmp_decending_prices);
+        }
+
+        let mut projection = DiGraph::new();
+        let token_nodes = (0..=max_token)
+            .map(|token_id| projection.add_node(token_id))
+            .collect::<Vec<_>>();
+        projection.extend_with_edges(orders.iter().flat_map(|(sell, orders)| {
+            let sell_node = token_nodes[*sell as usize];
+            orders.iter().map({
+                let token_nodes = &token_nodes;
```

Comment:
What exactly is this called? I didn't know you can do assignments inside the block but before the function definition starts?

Reply:
I just create a scope to help the compiler figure out lifetimes - the `token_nodes` borrow only lasts for the scope I defined.

Reply:
I think the term is a block expression.